readme.md
# dingtalk(钉钉) log4js appender

send log to ding bot.

ratelimit is 20/min

## Install
npm
```bash
npm install log4js @ruangnazi/log4js-dingbot
```
yarn
```bash
yarn add log4js @ruangnazi/log4js-dingbot
```

## Configuration

+ `type` - `@ruangnazi/log4js-dingbot`
+ `access_token` - `string` - your access_token
+ `secret` - `string` (optional, defaults to `null`) - your secret, see [robot Webhook api](https://developers.dingtalk.com/document/app/custom-robot-access)
+ `sendInterval` - `integer` (optional, defaults to `3`) - batch ding and send in ding every sendInterval seconds, if 0 then every log message will send an ding. see [api ratelimit](https://developers.dingtalk.com/document/app/invocation-frequency-limit)
+ `splitSendInterval` - `integer` (optional, defaults to `3`) - log message long than 19000 bytes will spilt to send every splitSendInterval seconds.
+ `shutdownTimeout` - `integer` (optional, defaults to `5`) - time in seconds to wait for emails to be sent during shutdown
+ `layout` - `object` (optional, defaults to basicLayout) - see [layouts](layouts.md)

## Example

```javascript
log4js.configure({
  appenders: {
    out: { type: 'stdout' },
    dingbot: {
      type: '@ruangnazi/log4js-dingbot',
      access_token: 'access_token',
      secret: 'secret',
      sendInterval: 5
    }
  },
  categories: { default: { appenders: ['out', 'dingbot'], level: 'debug' } }
})

const app = log4js.getLogger()
app.info('测试钉钉')
```

CHANGELOG.md
# @ruangnazi/log4js-dingbot changelog

# 1.0.1
+ split long message by toplimit 19000 bytes
+ change default sendInterval to 3s

test/demo.ts
/*
 * @Version: 0.0.1
 * @Author: ider
 * @Date: 2021-03-25 01:31:08
 * @LastEditors: ider
 * @LastEditTime: 2021-04-06 13:12:38
 * @Description:
 */

import log4js from 'log4js'

log4js.configure({
  appenders: {
    out: { type: 'stdout' },
    dingbot: {
      type: './src/index',
      access_token:
        '68c56f440c4cc6e894f0637b21418e8ab2594fd14a72e8228349c575c91375eb',
      secret:
        'SECefecc96347fe11df010a2a1a231d4bd521d891d9b74292d12af027b3ebaf5dcd',
      sendInterval: 1,
    }
  },
  categories: { default: { appenders: ['out', 'dingbot'], level: 'debug' } },
})


function generateRamStr(len) {
  // const chars = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789";
  const chars = '我们时常会遇到一些特定的场景需要我们生成随机字符串'
  let randomStr = "";
  for (let i = 0; i < len; i++) {
  randomStr += chars.charAt(Math.floor(Math.random() * chars.length));
  }
  return randomStr;
}


const app = log4js.getLogger()
const ss = generateRamStr(50)
console.log(unescape(encodeURIComponent(ss)).length)
console.log(new TextEncoder().encode(ss).length)
// app.info('测试钉钉')
app.info('测试钉钉1')
app.debug('测试钉钉2')
app.info('测试钉钉3')
app.info('测试钉钉4')
app.info('测试钉钉5')

app.info(ss)


test/tap/index-test.ts
/*
 * @Version: 0.0.1
 * @Author: ider
 * @Date: 2021-03-25 09:18:54
 * @LastEditors: ider
 * @LastEditTime: 2021-04-06 11:56:44
 * @Description:
 */

import sandbox from '@log4js-node/sandboxed-module'
import appender = require('../../src')
import { test } from 'tap'

function setupLogging(
  options,
  netError: Error | null,
  signError: string | null,
) {
  const fakeAxios = {
    msgs: [],
    url: null,
    data: null,
    params: null,
    post: async (url: string, data: any, params: object) => {
      fakeAxios.url = url
      fakeAxios.data = data
      fakeAxios.params = params
      fakeAxios.msgs.push(data.text.content)
      if (netError !== null) {
        throw netError
      } else if (signError !== null) {
        return {
          data: {
            errcode: 403,
            errmsg: signError,
          },
        }
      } else {
        return {
          data: { errcode: 0, errmsg: 'ok' },
        }
      }
    },
  }

  const fakeLayouts = {
    layout: function (type, config) {
      this.type = type
      this.config = config
      return evt => evt.data[0]
    },
    basicLayout: evt => evt.data[0],
    type: null,
    config: null,
  }

  const fakeConsole = {
    errors: [],
    error: function (msg, value) {
      this.errors.push({ msg: msg, value: value })
    },
  }

  options.type = '@ruangnazi/log4js-dingbot'
  const appenderModule = sandbox
    .require('../../src', {
      requires: {
        axios: fakeAxios,
      },
      globals: {
        console: fakeConsole,
        TextEncoder: TextEncoder
      },
    })
    .configure(options, fakeLayouts)

  return {
    appender: appenderModule,
    logger: msg => appenderModule({ data: [msg] }),
    axios: fakeAxios,
    layouts: fakeLayouts,
    console: fakeConsole,
  }
}

function checkMessages(assert, result) {
  result.axios.msgs.forEach((msg, i) => {
    assert.match(msg.trim(), `Log event #${i + 1}`)
  })
}

test('log4js dingbotAppender', batch => {
  batch.test('module should export a configure function', t => {
    t.type(appender.configure, 'function')
    t.end()
  })

  batch.test('dingbot setup', t => {
    const result = setupLogging(
      {
        access_token: 'TOKEN',
        secret: '#secret',
        sendInterval:0,
      },
      null,
      null,
    )

    result.logger('some log event')

    t.test('dingbot credentials should match', assert => {
      assert.equal(result.axios.params.params.access_token, 'TOKEN')
      assert.end()
    })
    t.end()
  })

  batch.test('basic usage', t => {
    const setup = setupLogging(
      {
        access_token: 'TOKEN',
        secret: '#secret',
        sendInterval: 0,
      },
      null,
      null,
    )

    setup.logger('Log event #1')

    t.equal(setup.axios.msgs.length, 1, 'should be one message only')
    checkMessages(t, setup)
    t.end()
  })

  batch.test('test no secret', t => {
    const setup = setupLogging(
      {
        access_token: 'TOKEN',
        sendInterval: 0,
      },
      null,
      null,
    )

    setup.logger('Log event #1')
    t.assert(setup.axios.params.params.sign === undefined, 'should be no sign')
    checkMessages(t, setup)
    t.end()
  })

  batch.test('nterval usage', async t => {
    const setup = setupLogging(
      {
        access_token: 'TOKEN',
        secret: '#secret',
        sendInterval: 1,
      },
      null,
      null,
    )
    setup.logger('Log event #1')
    setup.logger('Log event #2')
    setup.logger('Log event #3')
    await sleep(1100)
    t.equal(setup.axios.msgs.length, 1, 'should be one message only')
    checkMessages(t, setup)
    t.end()
  })

  batch.test('config with layout', t => {
    const result = setupLogging(
      {
        access_token: 'TOKEN',
        secret: '#secret',
        layout: {
          type: 'tester',
        },
      },
      null,
      null,
    )
    t.equal(result.layouts.type, 'tester', 'should configure layout')
    t.end()
  })

  batch.test('separate notification for each event', t => {
    const setup = setupLogging(
      {
        access_token: 'TOKEN',
        secret: '#secret',
        sendInterval: 0,
      },
      null,
      null,
    )
    setup.logger('Log event #1')
    setup.logger('Log event #2')
    setup.logger('Log event #3')

    t.equal(setup.axios.msgs.length, 3, 'should be three messages')
    checkMessages(t, setup)
    t.end()
  })

  batch.test('should send errors to console', async t => {
    const setup = setupLogging(
      {
        access_token: 'TOKEN',
        secret: '#secret',
        sendInterval: 0,
      },
      null,
      '444',
    )
    setup.logger('this will fail')

    await sleep(100)

    t.equal(setup.axios.msgs.length, 1, 'should be one message sent')
    t.equal(setup.console.errors.length, 1, 'should be one error on console')
    t.end()
  })

  batch.test('test net error', async t => {
    const setup = setupLogging(
      {
        access_token: 'TOKEN',
        secret: '#secret',
        sendInterval: 0,
      },
      Error('net error'),
      null,
    )
    setup.logger('this will fail')
    await sleep(100)
    t.equal(setup.axios.msgs.length, 1, 'should be one message sent')
    t.equal(setup.console.errors.length, 1, 'should be one error on console')
    t.equal(
      setup.console.errors[0].msg,
      'log4js:dingbot - Error sending log to dingbot: ',
    )
    t.end()
  })

  batch.test('should not break if shutdown ', t => {
    const setup = setupLogging(
      {
        access_token: 'TOKEN',
        secret: '#secret',
        sendInterval: 0,
        shutdownTimeout: 1,
      },
      null,
      null,
    )
    t.equal(setup.axios.msgs.length, 0, 'should not be any messages yet')

    setup.appender.shutdown(() => {
      t.equal(setup.axios.msgs.length, 0)
      t.end()
    })
  })

  batch.test(
    'should only wait a specified time for sending remaining msg',
    t => {
      const setup = setupLogging(
        {
          access_token: 'TOKEN',
          secret: '#secret',
          sendInterval: 0,
          shutdownTimeout: 1,
        },
        null,
        null,
      )
      setup.logger('Log event #1')
      t.equal(setup.axios.msgs.length, 1, 'should not be any messages yet')

      setup.appender.shutdown(() => {
        t.equal(setup.axios.msgs.length, 1)
        t.end()
      })
    },
  )

  batch.test('should send sendInterval', async t => {
    const setup = setupLogging(
      {
        access_token: 'TOKEN',
        secret: '#secret',
        sendInterval: 1,
      },
      null,
      null,
    )
    setup.logger('log event #0')
    setup.logger('log event #1')
    t.equal(setup.axios.msgs.length, 0, 'should not be any messages yet')
    await sleep(1500)
    t.equal(setup.axios.msgs.length, 1, 'should be one message sent')
    t.equal(setup.console.errors.length, 0, 'should be none error on console')
    t.end()
  })

  batch.end()
})

async function sleep(timeout: number) {
  return new Promise((resolve: Function) => {
    setTimeout(function () {
      resolve()
    }, timeout)
  })
}


test/sandbox-coverage.ts
/*
 * @Version: 0.0.1
 * @Author: ider
 * @Date: 2021-03-25 09:12:57
 * @LastEditors: ider
 * @LastEditTime: 2021-03-27 01:23:08
 * @Description:
 */

import NYC from 'nyc'
import sandbox from '@log4js-node/sandboxed-module'

sandbox.configure({
  sourceTransformers: {
    nyc: function (source) {
      if (this.filename.indexOf('node_modules') > -1) {
        return source
      }
      const nyc = new NYC()
      return nyc.instrumenter().instrumentSync(source, this.filename)
    },
  },
})


src/index.d.ts
/*
 * @Version: 0.0.1
 * @Author: ider
 * @Date: 2021-03-25 00:53:05
 * @LastEditors: ider
 * @LastEditTime: 2021-04-06 13:32:29
 * @Description:
 */

export interface DingbotAppender {
  type: 'dingbot'
  // your Dingding API token (see the slack docs)
  access_token: string
  // `string` (optional, defaults to `null`) - your secret, see [robot Webhook api](https://developers.dingtalk.com/document/app/custom-robot-access)
  secret?: string
  // `integer` (optional, defaults to `3`) - batch ding and send in ding every sendInterval seconds, if 0 then every log message will send an ding. see [api ratelimit](https://developers.dingtalk.com/document/app/invocation-frequency-limit)
  sendInterval?: number
  // `integer` (optional, defaults to `3`) - log message long than 19000 bytes will spilt to send every splitSendInterval seconds.
  splitSendInterval?: number
  // (defaults to 5) - time in seconds to wait for emails to be sent during shutdown
  shutdownTimeout?: number
  // (defaults to basicLayout) - the layout to use for the message.
  layout?: Function
}

export type Appender = DingbotAppender


src/index.ts
/*
 * @Version: 0.0.1
 * @Author: ider
 * @Date: 2021-03-25 00:44:34
 * @LastEditors: ider
 * @LastEditTime: 2021-04-06 13:06:47
 * @Description: 钉钉webhook通知 hook
 */

import axios from 'axios'
import crypto from 'crypto'
import { Appender } from './index.d'

function dingbotAppender(config: Appender, layout: Function) {
  // 超长日志的连续发送间隔，以 20/min 为参考
  const splitSendInterval =
    ('splitSendInterval' in config ? config.splitSendInterval : 3) * 1000
  // 发送间隔
  const sendInterval =
    ('sendInterval' in config ? config.sendInterval : 3) * 1000
  // 关闭前检查
  const shutdownTimeout =
    ('shutdownTimeout' in config ? config.shutdownTimeout : 5) * 1000
  const logEventBuffer = []
  // 长日子切割缓存
  let logLongLineTmp = ''
  let unsentCount = 0
  let sendTimer: NodeJS.Timeout

  function sendBuffer() {
    let body = ''
    let sentCount = 0
    // 缓存发送
    if (logLongLineTmp.length > 0) {
      if (logLongLineTmp.length > 6000) {
        body = logLongLineTmp.slice(0, 6000)
        logLongLineTmp = logLongLineTmp.slice(6000, -1)
        scheduleSend(splitSendInterval)
      } else {
        body = logLongLineTmp
        logLongLineTmp = ''
        // 安排发送事件中的日志
        scheduleSend()
      }
      sendDing(body, config)
    } else if (logEventBuffer.length > 0) {
      let messageLength = 0
      while (logEventBuffer.length > 0) {
        const currentLine = `${layout(logEventBuffer[0])}\n`
        // 判断是否超长,限制长度为 20000，超长就分布发送
        messageLength += new TextEncoder().encode(currentLine).length
        if (body.length > 0 && messageLength > 19000) {
          scheduleSend(splitSendInterval)
          break
        } else if (messageLength > 19000) {
          // 单条超过长度的,按照 6000 切割
          logEventBuffer.shift()
          logLongLineTmp = currentLine
          sendBuffer()
          return
        }
        logEventBuffer.shift()
        sentCount += 1
        body += currentLine
      }
      unsentCount -= sentCount
      sendDing(body, config)
    }
  }

  function scheduleSend(mysendInterval?: number) {
    const _sendInterval = mysendInterval ? mysendInterval : sendInterval
    if (!sendTimer) {
      sendTimer = setTimeout(() => {
        sendTimer = null
        sendBuffer()
      }, _sendInterval)
    }
  }

  function shutdown(cb: Function) {
    if (sendTimer) {
      clearTimeout(sendTimer)
    }
    sendBuffer()
    let timeout = shutdownTimeout
    ;(function checkDone() {
      if (unsentCount > 0 && timeout >= 0 && logLongLineTmp.length === 0) {
        timeout -= 100
        setTimeout(checkDone, 100)
      } else {
        cb()
      }
    })()
  }

  const appender = loggingEvent => {
    unsentCount += 1
    logEventBuffer.push(loggingEvent)
    if (sendInterval > 0) {
      scheduleSend()
    } else {
      sendBuffer()
    }
  }

  appender.shutdown = shutdown
  return appender
}

function configure(config, layouts) {
  let layout = layouts.basicLayout
  if (config.layout) {
    layout = layouts.layout(config.layout.type, config.layout)
  }
  return dingbotAppender(config, layout)
}

function sendDing(body: string, config: Appender) {
  const data = {
    msgtype: 'text',
    text: {
      content: body,
    },
  }
  const params: {
    access_token: string
    sign?: string
    timestamp?: number
  } = {
    access_token: config.access_token,
  }
  if (config.secret != null) {
    const timestamp = Date.now()
    params.sign = crypto
      .createHmac('sha256', config.secret)
      .update(`${timestamp}\n${config.secret}`)
      .digest('base64')
    params.timestamp = timestamp
  }
  axios
    .post(`https://oapi.dingtalk.com/robot/send`, data, {
      params,
    })
    .then(response => {
      if (response.data?.errcode !== 0) {
        console.error(
          'log4js:dingbot - Error sending log to dingbot: ',
          response.data,
        )
      }
    })
    .catch(err => {
      console.error('log4js:dingbot - Error sending log to dingbot: ', err)
    })
}

export { configure }
